bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,193 @@
+use std::{collections::HashMap, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use once_cell::sync::Lazy;
+use pingora::{
+    http::ResponseHeader,
+    lb::Backend,
+    prelude::*,
+    upstreams::peer::{Peer, ALPN},
+};
+use pingora_core::{upstreams::peer::HttpPeer, Result};
+use pingora_limits::rate::Rate;
+use pingora_proxy::{ProxyHttp, Session};
+use tokio::sync::RwLock;
+use tracing::error;
+
+/// Timeout for the requests
+const TIMEOUT_SECS: Option<Duration> = Some(Duration::from_secs(100));
+/// Maximum number of items per queue
+const MAX_QUEUE_ITEMS: usize = 10;
+
+/// Load balancer that uses a round robin strategy
+pub struct LB(Arc<LoadBalancer<RoundRobin>>);
+
+impl LB {
+    pub fn new(upstreams: LoadBalancer<RoundRobin>) -> Self {
+        Self(Arc::new(upstreams))
+    }
+}
+
+/// Rate limiter
+static RATE_LIMITER: Lazy<Rate> = Lazy::new(|| Rate::new(Duration::from_secs(1)));
+
+/// Maximum amount of request per second per client
+static MAX_REQ_PER_SEC: isize = 5;
+
+/// Shared state. It is a map of backends to a vector of request IDs
+static QUEUES: Lazy<RwLock<HashMap<Backend, Vec<String>>>> =
+    Lazy::new(|| RwLock::new(HashMap::new()));
+
+#[async_trait]
+impl ProxyHttp for LB {
+    type CTX = ();
+
+    fn new_ctx(&self) {}
+
+    async fn upstream_peer(
+        &self,
+        session: &mut Session,
+        _ctx: &mut Self::CTX,
+    ) -> Result<Box<HttpPeer>> {
+        let upstream = self.0.select(b"", 256).unwrap();
+
+        // read request ID from headers
+        let request_id = session
+            .get_header("X-Request-ID")
+            .expect("Request ID not found")
+            .to_str()
+            .expect("Invalid header value");
+
+        {
+            let mut ctx_guard = QUEUES.write().await;
+            let backend_queue = ctx_guard.entry(upstream.clone()).or_insert_with(Vec::new);
+
+            // Limit queue length to MAX_QUEUE_ITEMS requests
+            if backend_queue.len() >= MAX_QUEUE_ITEMS {
+                panic!("Too many requests in the queue");
+                // return Err(Box::new("Too many requests in the queue".into()));
+            }
+
+            backend_queue.push(request_id.to_string());
+        }
+
+        // Wait for the request to be at the front of the queue
+        loop {
+            // We use a new scope for each iteration to release the lock
+            {
+                let ctx_guard = QUEUES.read().await;
+                if let Some(backend_queue) = ctx_guard.get(&upstream) {
+                    if backend_queue[0] == request_id {
+                        break;
+                    }
+                } else {
+                    // TODO: replace this panic with an error
+                    panic!("Upstream not found");
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
```

Comment:
Maybe it's overkill for this context, but I think a [conditional variable](https://doc.rust-lang.org/std/sync/struct.Condvar.html) would be perfect here.

Reply:
I just realized we are in a tokio context here, I think [this](https://docs.rs/tokio/latest/tokio/sync/struct.Notify.html) would be the alternative.

Reply:
This sounds like a good alternative, let me investigate more on this